### PR TITLE
backend: implement basic functionality (PROJQUAY-7076)

### PR DESCRIPTION
**Pull request by Sunandadadi**

### CODE REVIEW

**Comment by Sunandadadi** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/quay/quay/2895?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2895** <a href="https://app.graphite.dev/github/pr/quay/quay/2895?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#2894** <a href="https://app.graphite.dev/github/pr/quay/quay/2894?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Sunandadadi and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: util/notification.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                    RepositoryNotification.event_config_json,
+                )
+                .where(
+                    RepositoryNotification.event == event.id,
+                    RepositoryNotification.number_of_failures < 3,
+                    (
+                        RepositoryNotification.last_ran_ms
+                        < get_epoch_timestamp_ms() - task_run_interval_ms
+                    )
+                    | (RepositoryNotification.last_ran_ms.is_null(True)),
```
Comment (syed) [inline util/notification.py:45]: make sure `last_ran_ms` is indexed

Comment (Sunandadadi) [reply util/notification.py:45]: yes, added in index [here](https://github.com/quay/quay/pull/2894/files#diff-884d5b5805e4e80294281e5415112e55f65fe5c96a1471e04d190120b73d5d5fR25-R30)

File: endpoints/api/repositorynotification_models_interface.py [hunk 2]
```diff
@@ -9,6 +9,7 @@ class RepositoryNotification(
     namedtuple(
         "RepositoryNotification",
         [
```
Comment (syed) [inline endpoints/api/repositorynotification_models_interface.py:11]: is this the DB ID? we shouldn't expose this to an external API

File: workers/gc/gcworker.py [hunk 3]
```diff
@@ -30,6 +31,13 @@
         self.add_operation(
             self._garbage_collection_repos, app.config.get("GARBAGE_COLLECTION_FREQUENCY", 30)
         )
+        self.add_operation(
+            self._scan_notifications, app.config.get("GARBAGE_COLLECTION_FREQUENCY", 30)
+        )
+
+    def _scan_notifications(self):
+        # scan for tags that are expiring based on configured RepositoryNotifications
+        scan_for_image_expiry_notifications(event_name="repo_image_expiry")
```
Comment (syed) [inline workers/gc/gcworker.py:40]: can you use `RepoImageExpiryEvent.getname()`

File: endpoints/api/repositorynotification_models_interface.py (hunk 2 above)
Comment (Sunandadadi) [reply endpoints/api/repositorynotification_models_interface.py]: yes, good point, I removed it

**Review by syed — APPROVED** [review]
Lgtm

**Pull request merged**